haystack/components/extractors/llm_metadata_extractor.py

```diff
@@ -311,52 +294,10 @@ def from_dict(cls, data: Dict[str, Any]) -> "LLMMetadataExtractor":
             An instance of the component.
         """
 
-        init_parameters = data.get("init_parameters", {})
-
-        # new deserialization with chat_generator
-        if init_parameters.get("chat_generator") is not None:
-            chat_generator_class = import_class_by_name(data["init_parameters"]["chat_generator"]["type"])
-            assert hasattr(chat_generator_class, "from_dict")  # we know but mypy doesn't
-            chat_generator_instance = chat_generator_class.from_dict(init_parameters["chat_generator"])
-            data["init_parameters"]["chat_generator"] = chat_generator_instance
-            return default_from_dict(cls, data)
-
-        # legacy deserialization
-        if "generator_api" in init_parameters:
-            data["init_parameters"]["generator_api"] = LLMProvider.from_str(data["init_parameters"]["generator_api"])
-
-        if "generator_api_params" in init_parameters:
-            # Check all the keys that need to be deserialized
-            azure_openai_keys = ["azure_ad_token"]
-            aws_bedrock_keys = [
-                "aws_access_key_id",
-                "aws_secret_access_key",
-                "aws_session_token",
-                "aws_region_name",
-                "aws_profile_name",
-            ]
-            deserialize_secrets_inplace(
-                data["init_parameters"]["generator_api_params"], keys=["api_key"] + azure_openai_keys + aws_bedrock_keys
-            )
-
-            # For VertexAI
-            if "generation_config" in init_parameters["generator_api_params"]:
-                data["init_parameters"]["generation_config"] = GenerationConfig.from_dict(
-                    init_parameters["generator_api_params"]["generation_config"]
-                )
-
-            # For AzureOpenAI
-            serialized_azure_ad_token_provider = init_parameters["generator_api_params"].get("azure_ad_token_provider")
-            if serialized_azure_ad_token_provider:
-                data["init_parameters"]["azure_ad_token_provider"] = deserialize_callable(
-                    serialized_azure_ad_token_provider
-                )
-
-            # For all
-            serialized_callback_handler = init_parameters["generator_api_params"].get("streaming_callback")
-            if serialized_callback_handler:
-                data["init_parameters"]["streaming_callback"] = deserialize_callable(serialized_callback_handler)
-
+        chat_generator_class = import_class_by_name(data["init_parameters"]["chat_generator"]["type"])
```

Comment:
Now we are assuming that `chat_generator` is always set when calling `from_dict`. However, what if a user serialized the pipeline with 2.11 and deserializes with 2.12? We should support that too.

Reply:
I think you could actually keep the `from_dict` as is if the init method handles using the ChatGenerator even if a `generator_api` parameter value was provided.

Reply:
right! https://github.com/deepset-ai/haystack/pull/9099/commits/f026f52f9b7cbcecad4f2a90e3a1ba9f121e82bc